kernel/sched.c

```diff
@@ -632,10 +618,7 @@ void z_sched_wake_thread(struct k_thread *thread, bool is_timeout)
 			if (thread->base.pended_on != NULL) {
 				unpend_thread_no_timeout(thread);
 			}
-			z_mark_thread_as_started(thread);
-			if (is_timeout) {
-				z_mark_thread_as_not_suspended(thread);
-			}
```

Comment:
I think (for now) we want ...
```
    if (is_timeout) {
        z_mark_thread_as_not_suspend(thread);
    }
```
z_sched_wake_thread() can be called from more than just the timer and we don't necessarily want to resume/unsuspend the thread.

Reply:
Everything else looks great!

Reply:
When would we not? I mean, obviously with your SLEEPING patch we'd care about the distinction (and would have to add code here to handle it). But the code as it stands doesn't have that and "wake" and "resume" are synonyms. What's the circumstance where you'd call "wake thread" and expect it not to run?

Reply:
At the present time, z_sched_wake_thread() is also called from the events code with `is_timeout = false` to wake a pended thread. Unconditionally resuming/unsuspending the thread creates an inconsistency when unpending suspended threads that were waiting for a resource. In the case of message queues and semaphores (which do not use z_sched_wake_thread()) the thread may gain the resource but still remain suspended. In the case of events, the suspended thread is resumed/unsuspended upon matching the events.

Reply:
> Unconditionally resuming/unsuspending the thread creates an inconsistency when unpending suspended threads that were waiting for a resource.

Right, but you've literally got a patch in the pipeline that is trying to address this inconsistency. :)  Why demand it be fixed ahead of time when we know the disconnect is deeper?

I guess I'm confused: does this change existing behavior or no?  If it breaks something (especially something with a test) it should be fixed.  If it's just us arguing about scheduler philosophy, then my vote is to merge it like this (which seems clearer and safer to me) and then you can come by after this merges and fix it correctly.

Reply:
Just to repeat again, because we keep hitting the same disconnect: you and I have **different mental models of what "suspend" is supposed to mean**.  To me, and (most of) the code, "suspend" is a **verb** that means "stop the thread immediately and make sure no further kernel work for it occurs, you can start it again by any means".

You want it to be an **adjective** that means "thread will not run for any reason until resume is called".  And I don't even think that's the wrong design (though it does have size costs, c.f. the extra state bit we're here trying to make room for).  But the code here is AFAICT in keeping with the existing design.  If we're going to change it we should do it in one atomic PR, not piecewise.